src/qililab/instruments/quantum_machines/quantum_machines_cluster.py

```diff
@@ -0,0 +1,354 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Quantum Machines Manager class."""
+import os
+from dataclasses import dataclass
+from typing import Any, Dict
+
+import numpy as np
+from qm import DictQuaConfig, QuantumMachine, QuantumMachinesManager, SimulationConfig
+from qm.jobs.running_qm_job import RunningQmJob
+from qm.octave import QmOctaveConfig
+from qm.qua import Program
+
+from qililab.instruments.instrument import Instrument
+from qililab.instruments.utils import InstrumentFactory
+from qililab.typings import InstrumentName, Parameter, QMMDriver
+from qililab.utils import merge_dictionaries
+
+
+@InstrumentFactory.register
+class QuantumMachinesCluster(Instrument):
+    """Class defining the Qililab Quantum Machines Cluster instrument in Qililab.
+
+    This class allows Qililab control and communication with an instance of the
+    Quantum Machines Manager, which is the central class to interact with Quantum Machines instruments.
+    The manager is responsible, through the use of QUA sequences, of setting the Quantum Machines instruments
+    in the right manner for Quantum Control.
+
+    Args:
+        name (InstrumentName): Name of the Instrument.
+        device (QMMDriver): Instance of the Quantum Machines Manager Driver class.
+        settings (QMMSettings): Settings of the instrument.
+    """
+
+    name = InstrumentName.QUANTUM_MACHINES_CLUSTER
+
+    @dataclass
+    class QuantumMachinesClusterSettings(Instrument.InstrumentSettings):
+        """Settings for Quantum Machines Cluster instrument.
+
+        Args:
+            address (str): IP address to connect to the Quantum Machines instruments.
+            cluster (str): Name of the cluster to connect to.
+            num_controllers (int): Number of controllers (instruments) in the quantum machines stack.
+            octaves (list[dict[str, Any]]): List of octaves the quantum machines stack has.
+            controllers (list[dict[str, Any]]): List of controllers (instruments) the quantum machines stack has.
+            elements (list[dict[str, Any]]): List of elements (buses) the quantum machines stack has.
+        """
+
+        address: str
+        cluster: str
+        run_octave_calibration: bool
+        octaves: list[dict[str, Any]]
+        controllers: list[dict[str, Any]]
+        elements: list[dict[str, Any]]
+
+        def to_qua_config(self) -> Dict[str, Any]:
+            """Creates the Quantum Machines QUA config dictionary.
+
+            Creates, an instance of a dictionary in the format that QuantumMachines expects the config dictionary to be.
```

Comment:
We need to expand on this docstring to include mixers and octaves. Maybe modify slightly that the empty fields are the ones that will be fulfilled by the QProgram dictionary configuration.

Reply:
Done